RFD: Kube MFA sessions

AntonAM opened the pull request:
This RFD describes extension of idea of DB MFA Sessions for Kubernetes access to improve user experience.

[Rendered version](https://github.com/gravitational/teleport/blob/b2a81806728c0db308ee27f7f761160cb23d047b/rfd/0xxx-kube-mfa-sessions.md)

Relevant issue: https://github.com/gravitational/teleport/issues/7568

AntonAM committed:
Add RFD for kube MFA sessions

zmb3 commented:
In the future, please try to follow the branch naming conventions for RFDs: https://github.com/gravitational/teleport/blob/785fa046276cee7adde54a7c5cc31b013cc7a65f/rfd/0000-rfds.md?plain=1#L68-L70

tigrato commented on rfd/0xxx-kube-mfa-sessions.md:
```diff
@@ -0,0 +1,142 @@
+Local proxy will also reissue certificates and ask for MFA check/credentials if required.
+
+Unlike DB MFA sessions, Kubernetes local proxy will only have one mode of operation (DB access has tunnelled and not tunnelled mode). 
+Since the pattern of usage for Kubernetes is mostly short-lived commands, analogue of tunneled mode for the local proxy makes the most sense.
```
Didn't @greedy52 already introduced the non-tunneled mode? 

If so, we should differentiate the behavior and by default disable the MFA caching.
Kube local proxy is still required when running behind an L7 load balancer and we don't want to enable MFA caching by default when running behind L7 load balancers.

If users want MFA caching, they must explicitly enable it.
This changes nothing in terms of code, it's just an extra flag.

It probably requires some UX to inform kubectl when a certificate must be reissued

greedy52 replied:
> analogue of tunneled mode for the local proxy makes the most sense

ya. The definition of tunnelled vs non-tunnelled for DB is probably different than Kube. Tunnelled for DB means database client sends data to local proxy in plain database protocol and local proxy will take care of the mTLS part, where non-tunnelled means database client sends encrypted database with Teleport-issued cert.

I would say `tsh proxy kube` is closer to tunnelled mode since kube clients are not using Teleport-issued cert against local proxy, and local proxy is handling the Teleport-issued cert.

AntonAM replied:
Yep, as @greedy52 said, what was currently added is already I would say tunneled mode. I think having only this mode makes sense for Kube.

tigrato reviewed (commented):
LGTM

What's the UX when renewing MFA cached certificates? 
- Will you return an error to kubectl? 
- present the prompt in the `tsh proxy kube` terminal 
- extend our credentials plugin to help the MFA handling using kubectl when running a local proxy without requiring the user to switch terminals.

greedy52 commented on rfd/0xxx-kube-mfa-sessions.md:
```diff
@@ -0,0 +1,142 @@
+--------------------- ----------------------
+my-teleport.com       test-kube-cluster
+
+Enter an OTP code from a device:
```
The current implementation of kube local proxy supports multiple kube clusters. It's convenient if the user wants to log in to multiple clusters at the same time but I believe the MFA request is per kube cluster as well.  Are we prompting multiple times in that case?

AntonAM commented:
@tigrato 
As per discussion, added section mentioning UX - we require input in the `tsh proxy kube` terminal, but will return an error after relatively short timeout. And added future development section.

tigrato commented on rfd/0xxx-kube-mfa-sessions.md:
```diff
@@ -0,0 +1,152 @@
+ it will be automatically picked up.
+In addition users can utilize aliases to make it even simpler to invoke Kubernetes commands with running local proxy. 
+
+In case of relogin user might not notice immediately that running proxy requires input. To improve the situation we'll add 
+a timeout for relogin operation and return an error to the user's terminal with kubectl, so they can understand that local proxy
+requires attention.
```
```suggestion
In the case of relogin user might not notice immediately that running proxy requires input. To improve the situation we'll add 
a timeout for the request that triggered the relogin operation and return an error to kubectl if the login didn't happen until the timeout occurs, so they can understand that the local proxy
requires attention.
```
